src/algorithms/reco/ScatteredElectronsEMinusPz.cc

```diff
@@ -0,0 +1,154 @@
+#include <edm4hep/Vector3f.h>
+#include <edm4hep/utils/vector_utils.h>
+#include <fmt/core.h>
+#include <podio/ObjectID.h>
+#include <cmath>
+#include <gsl/pointers>
+#include <vector>
+#include <TLorentzVector.h>
+
+#include "Beam.h"
+#include "ScatteredElectronsEMinusPz.h"
+
+using ROOT::Math::PxPyPzEVector;
+using ROOT::Math::PxPyPzMVector;
+
+namespace eicrecon {
+
+  /**
+   * @brief Initialize the ScatteredElectronsEMinusPz Algorithm
+   *
+   * @param logger
+   */
+  void ScatteredElectronsEMinusPz::init(std::shared_ptr<spdlog::logger>& logger) {
+    m_log = logger;
+  }
+
+  /**
+   * @brief Produce a list of scattered electron candidates
+   *
+   * @param rcparts - input collection of all reconstructed particles
+   * @param rcele  - input collection of all electron candidates
+   * @return std::unique_ptr<edm4eic::ReconstructedParticleCollection>
+   */
+  std::unique_ptr<edm4eic::ReconstructedParticleCollection> ScatteredElectronsEMinusPz::execute(
+                const edm4eic::ReconstructedParticleCollection *rcparts,
+                const edm4eic::ReconstructedParticleCollection *rcele
+        ){
+
+    // this map will store intermediate results
+    // so that we can sort them before filling output
+    // collection
+    std::map<double, edm4eic::MutableReconstructedParticle> scatteredElectronsMap;
+
+    // our output collection of scattered electrons
+    // ordered by E-Pz
+    auto out_electrons =  std::make_unique<
+        edm4eic::ReconstructedParticleCollection
+      >();
+
+    m_log->trace( "We have {} candidate electrons",
+        rcele->size()
+      );
+
+    // Lorentz Vector for the scattered electron,
+    // hadronic final state, and individual hadron
+    // We do it here to avoid creating objects inside the loops
+    PxPyPzMVector  vScatteredElectron;
+    PxPyPzMVector  vHadronicFinalState;
+    PxPyPzMVector  vHadron;
+
+    for ( const auto& e: *rcele ) {
+      // Do not cut on charge to account for charge-symmetric background
+
+      // reset the HadronicFinalState
+      vHadronicFinalState.SetCoordinates(0, 0, 0, 0);
+
+      // Set a vector for the electron we are considering now
+      vScatteredElectron.SetCoordinates(
+          e.getMomentum().x,
+          e.getMomentum().y,
+          e.getMomentum().z,
+          m_electron
+        );
+
+      // Loop over reconstructed particles to
+      // sum hadronic final state
+      for (const auto& p: *rcparts) {
+        // this is a hack - getObjectID() only works within
+        // a collections, not unique across all collections.
+        // What we want is to add all reconstructed particles
+        // except the one we are currently considering as the
+        // (scattered) electron candidate.
+        // This does work though and in general it has only
+        // one match as I would hope (tested on pythia events)
+        if (abs( edm4hep::utils::magnitude(p.getMomentum()) - edm4hep::utils::magnitude(e.getMomentum()) ) > 0.01 )
```

Comment:
Constants need units when compared with unit-full quantities.
```suggestion
        if (abs( edm4hep::utils::magnitude(p.getMomentum()) - edm4hep::utils::magnitude(e.getMomentum()) ) > 0.01 * dd4hep::GeV )
```
Or we should use `std::numeric_limits<double>::epsilon()`.